db/view/view.cc

```diff
@@ -1797,6 +1814,13 @@ get_view_natural_endpoint(
         }
     }
 
+    if (rack_aware_pairing && !simple_rack_aware_pairing) {
+        // For the complex rack_aware_pairing case, nodes are already filtered by datacenter
+        // so just stable_sort by rack (and keep ordinal order within the racks)
+        std::ranges::stable_sort(base_endpoints, std::ranges::less{}, std::mem_fn(&locator::node::rack));
+        std::ranges::stable_sort(view_endpoints, std::ranges::less{}, std::mem_fn(&locator::node::rack));
```

Comment:
I'm not sure the non-simple rack aware pairing helps much. If we have base replicas on racks R1,R2,R3 and view replicas on R2,R3,R4 (already sorted), we'll get pairings R1->R2, R2->R3, R3->R4 while we could've paired replicas on 2 racks - by offsetting the count of replicas on one rack we may prevent many more same-rack pairings.

Perhaps we should do a similar thing as in the self-pairing optimization (which is actually similar to this optimization), where we would pair base and view replicas from the same rack (as many as we can) first, remove them from the list, and then pair the remaining replicas according to the lists?

Reply:
I had a version that tried a best match in the first pass, matching pairs in all racks based on the minimum number of replicas in both base and view, while matching all the other replicas across racks by their ordinal order.

@raphaelsc didn't like that and asked to use sorting instead.
The best matching algorithm still used a sort by oridinal number, so it has a complexity of n*log(n) as well, but typically for a smaller N, and it is needed in a scenario that is presumed to be rare (rf != number of racks), at least in the cloud.  so I'm not sure how much it's worth the effort.

@avikivity do you have an opinion about this?

Reply:
See https://github.com/bhalevy/scylla/commit/722ad7c0ebfd026ac93e00e735065fb7e3a2eabf